internal/impl/redpanda/processor_data_transforms.go

```diff
@@ -0,0 +1,414 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redpanda
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+	"github.com/tetratelabs/wazero/sys"
+
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+func dataTransformProcessorConfig() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Categories("Redpanda").
+		Summary("Executes a Redpanda Data Transform as a processor").
+		Description(`
+This processor executes a Redpanda Data Transform WebAssembly module, calling OnRecordWritten for each message being processed.
+`).
+		Field(service.NewStringField("module_path").
+			Description("The path of the target WASM module to execute.")).
+		Field(service.NewInterpolatedStringField("input_key").
+			Description("An optional key to populate for each message.").Optional()).
+		Field(service.NewStringField("output_key").
+			Description("An optional name of metadata for an output message key.").Optional()).
+		Field(service.NewMetadataFilterField("input_headers").
+			Description("Determine which (if any) metadata values should be added to messages as headers.").
+			Optional()).
+		Field(service.NewMetadataFilterField("output_metadata").
+			Description("Determine which (if any) message headers should be added to the output as metadata.").
+			Optional()).
+		Field(service.NewInterpolatedStringField("timestamp").
+			Description("An optional timestamp to set for each message. When left empty, the current timestamp is used.").
+			Example(`${! timestamp_unix() }`).
+			Example(`${! metadata("kafka_timestamp_unix") }`).
+			Optional().
+			Advanced()).
+		Field(service.NewDurationField("timeout").
+			Description("The maximum period of time for a message to be processed").
+			Default("10s").
+			Advanced()).
+		Field(service.NewIntField("max_memory_pages").
+			Description("The maximum amount of wasm memory pages (64KiB) that an individual wasm module instance can use").
+			Default("1600"). // 100MiB
```

Comment:
Does 1600 correspond to 100MiB?

Reply:
Yes. 1600 pages and each page is 64KiB so it's 100MiB of memory. It's enough to be dangerous in Wasm but not too much to overwhelm the system (hopefully).

Reply:
Done!